src/main/scala/tilelink/SRAM.scala

```diff
@@ -173,73 +219,107 @@ class TLRAM(
-      d_lanes     := Cat(a_lanes.reverse)
-      when (!a_read && (a_sublane || a_atomic)) {
-        d_rmw_mask := in.a.bits.mask
-        d_rmw_data := in.a.bits.data
-      }
-      d_held_error:= Bool(false)
+      r_full     := true.B
+      r_sublane  := a_sublane
+      r_opcode   := in.a.bits.opcode
+      r_param    := in.a.bits.param
+      r_size     := in.a.bits.size
+      r_source   := in.a.bits.source
+      r_read     := a_read
+      r_atomic   := a_atomic
+      r_sublane  := a_sublane
+      r_address  := in.a.bits.address
+      r_poison   := in.a.bits.corrupt
+      r_mask     := in.a.bits.mask
+      when (!a_read) { r_rmw_data := in.a.bits.data }
     }
 
+    // Split data into eccBytes-sized chunks:
+    val a_data = VecInit(Seq.tabulate(lanes) { i => in.a.bits.data(eccBytes*8*(i+1)-1, eccBytes*8*i) })
+    val r_data = VecInit(Seq.tabulate(lanes) { i => r_rmw_data(eccBytes*8*(i+1)-1, eccBytes*8*i) })
+    val d_data = VecInit(Seq.tabulate(lanes) { i => d_updated(8*eccBytes*(i+1)-1, 8*eccBytes*i) })
+
+    // Which data chunks get poisoned
+    val a_poisonv = VecInit(Seq.fill(lanes) { in.a.bits.corrupt })
+    val r_poisonv = VecInit(Seq.fill(lanes) { r_poison })
+    val d_poisonv = VecInit(Seq.tabulate(lanes) { i =>
+      val upd = d_mask(eccBytes*(i+1)-1, eccBytes*i)
+      (!upd.andR && d_uncorrectable(i)) || d_poison // sub-lane writes should not correct uncorrectable
+    })
+
+    val a_lanes = Cat(Seq.tabulate(lanes) { i => in.a.bits.mask(eccBytes*(i+1)-1, eccBytes*i).orR }.reverse)
+
     // SRAM arbitration
     val a_fire = in.a.fire()
-    val wen =  d_wb || (a_fire && !a_ren)
-//  val ren = !d_wb && (a_fire &&  a_ren)
-    val ren = !wen && a_fire // help Chisel infer a RW-port
-
-    val addr   = Mux(d_wb, d_address, a_address)
-    val sel    = Mux(d_wb, Vec(d_wb_lanes), Vec(a_lanes))
-    val dat    = Mux(d_wb, d_wb_data, a_data)
-    val poison = Mux(d_wb, Vec(d_wb_poison), Vec.fill(lanes) { in.a.bits.corrupt })
-    val coded  = Vec((dat zip poison) map { case (d, p) =>
+    val a_ren = a_read || a_atomic || a_sublane
+    val r_ren = r_read || r_atomic || r_sublane
+    val wen = d_wb || Mux(r_replay, !r_ren, a_fire && !a_ren)
+    val ren = !wen && (a_fire || r_replay) // help Chisel infer a RW-port
+
+    val addr   = Mux(d_wb, d_address, Mux(r_replay, r_address, in.a.bits.address))
+    val sel    = Mux(d_wb, d_lanes,   Mux(r_replay, r_lanes,   a_lanes))
+    val dat    = Mux(d_wb, d_data,    Mux(r_replay, r_data,    a_data))
+    val poison = Mux(d_wb, d_poisonv, Mux(r_replay, r_poisonv, a_poisonv))
+    val coded  = VecInit((dat zip poison) map { case (d, p) =>
       if (code.canDetect) code.encode(d, p) else code.encode(d)
     })
 
-    d_raw_data := mem.read(addr, ren)
-    when (wen) { mem.write(addr, coded, sel) }
+    val index = Cat(mask.zip((addr >> log2Ceil(beatBytes)).asBools).filter(_._1).map(_._2).reverse)
+    r_raw_data := mem.read(index, ren) holdUnless RegNext(ren)
```

Comment:
Not that it matters much, but a while back I defined `mem.readAndHold(index, ren)` to summarize this pattern